File: common_book/src/main/java/com/intelligent/reader/activity/DownloadManagerActivity.java
package com.intelligent.reader.activity;

import com.intelligent.reader.BuildConfig;
import com.intelligent.reader.R;
import com.intelligent.reader.app.BookApplication;
import com.intelligent.reader.read.help.BookHelper;
import com.intelligent.reader.receiver.DownBookClickReceiver;
import com.intelligent.reader.view.DownloadPager;

import net.lzbook.kit.app.BaseBookApplication;
import net.lzbook.kit.appender_loghub.StartLogClickUtil;
import net.lzbook.kit.book.component.service.DownloadService;
import net.lzbook.kit.book.view.MyDialog;
import net.lzbook.kit.data.bean.Book;
import net.lzbook.kit.data.db.BookDaoHelper;
import net.lzbook.kit.utils.AppUtils;
import net.lzbook.kit.utils.FrameBookHelper;
import net.lzbook.kit.utils.FrameBookHelper.MultiComparator;
import net.lzbook.kit.utils.StatServiceUtils;

import android.app.Activity;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.AdapterView.OnItemLongClickListener;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * 下载管理
 */
public class DownloadManagerActivity extends BaseCacheableActivity implements OnClickListener, OnItemClickListener, OnItemLongClickListener, DownloadService.OnDownloadListener {
    private static final String TAG = "DownloadManagerActivity";
    public DownloadPager views;
    public FrameBookHelper.MultiComparator multiComparator;
    public long lastClickTime;
    private ImageView back_btn;
    private TextView title_name_btn;
    private TextView editBtn;
    private ViewGroup content_layout;
    private ArrayList<Book> downloadingBooks;
    private DownloadService downloadService;
    private BookDaoHelper mBookDaoHelper;
    private Book adpBook;
    private MyDialog alertDialog;
    //private ProgressDialog dialog;
    private ServiceConnection sc = new ServiceConnection() {

        @Override
        public void onServiceDisconnected(ComponentName name) {
//            if (dialog != null) {
//                dialog.dismiss();
//            }
        }

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            downloadService = ((DownloadService.MyBinder) service).getService();
            BaseBookApplication.setDownloadService(downloadService);
            downloadService.setUiContext(getApplicationContext());
            downloadService.setOnDownloadListener(DownloadManagerActivity.this);
//            if (dialog != null) {
//                dialog.dismiss();
            getDownLoadBookList(false);
//            }

        }
    };

    @Override
    public void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);
        try {
            setContentView(R.layout.download_manager);
        } catch (Resources.NotFoundException e) {
            e.printStackTrace();
        }
        downloadService = BaseBookApplication.getDownloadService();
        multiComparator = new MultiComparator();
        initView();
        initData();
        initListener();
    }

    @Override
    protected void onStart() {
        super.onStart();
        initService();
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    @Override
    protected void onDestroy() {

        if (alertDialog != null) {
            alertDialog = null;
        }

        try {
            setContentView(R.layout.empty);
        } catch (Resources.NotFoundException e) {
            e.printStackTrace();
        }

        super.onDestroy();
        if (BuildConfig.DEBUG) {
            BookApplication.getRefWatcher().watch(this);
        }
    }

    @Override
    protected void onResume() {
        try {
            super.onResume();
            freshBooks(false);
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    /**
     * 初始化离线下载服务
     */
    private void initService() {
    }

    public void stopDownloadbook(String book_id) {
        if (downloadService != null) {
            downloadService.cancelTask(book_id);
        }
    }

    private void initView() {
        // title初始化
        back_btn = (ImageView) findViewById(R.id.title_back_btn);
        editBtn = (TextView) findViewById(R.id.btn_edit);
        content_layout = (ViewGroup) findViewById(R.id.content_layout);
        title_name_btn = (TextView) findViewById(R.id.title_name_btn);
    }

    private void initListener() {
        back_btn.setOnClickListener(this);
        editBtn.setOnClickListener(this);

        title_name_btn.setOnClickListener(this);
        views.setDeleteItemListener(new DownloadPager.DeleteItemListener() {
            @Override
            public void onSuccess() {
                if (views.booksData.size() == 0) {
                    editBtn.setVisibility(View.GONE);
                } else {
                    editBtn.setText("编辑");
                }
            }

            @Override
            public void onFailed() {
                editBtn.setText("编辑");
            }
        });
        views.getListView().setOnItemClickListener(this);
        views.getListView().setOnItemLongClickListener(this);
        if (downloadService == null) {
            reStartDownloadService(this);
            downloadService = BaseBookApplication.getDownloadService();
        } else {
            downloadService.setOnDownloadListener(this);
            downloadService.setUiContext(getApplicationContext());
        }
    }

    private void initData() {
        mBookDaoHelper = BookDaoHelper.getInstance();
        downloadingBooks = new ArrayList<>();
        views = new DownloadPager(getApplicationContext(), this, downloadingBooks);
        content_layout.addView(views);
    }

    /**
     * 检测书籍下载状态分类，并刷新ui
     */
    public void freshBooks(boolean hasDeleted) {
        if (downloadService == null) {
            reStartDownloadService(this);
            return;
        }

        getDownLoadBookList(hasDeleted);
    }

    private void getDownLoadBookList(boolean hasDeleted) {
        if (mBookDaoHelper == null) {
            mBookDaoHelper = BookDaoHelper.getInstance();
        }
        ArrayList<Book> books = mBookDaoHelper.getBooksOnLineList();
//        Log.e(TAG, "addBookToService:"+books);
        addBookToService(books);
        // TODO 从service获取books列表 FIXME
        if (downloadService != null) {
//            Log.e(TAG, "getDownloadTaskNoFinish:"+downBooks);
            if (downloadingBooks != null && views != null) {
                downloadingBooks.clear();
                downloadingBooks.addAll(books);
                Collections.sort(downloadingBooks);
                Collections.sort(downloadingBooks, multiComparator);

                if (downloadingBooks.size() == 0) {
                    editBtn.setVisibility(View.GONE);
                } else {
                    if (AppUtils.getPackageName().equals("cc.kdqbxs.reader")) {
                        editBtn.setVisibility(View.GONE);
                    } else {
                        editBtn.setVisibility(View.VISIBLE);
                        if (hasDeleted) {
                            editBtn.setText("编辑");
                        }
                    }
                }
                views.freshBookList(downloadingBooks);
            }
        } else {
            Log.e(TAG, "downloadService == null");
        }
    }

    private void reStartDownloadService(Activity context) {
        //dialog = ProgressDialog.show(context, "", "重启服务中....", true, true);
        Intent intent = new Intent();
        intent.setClass(context, DownloadService.class);
        context.startService(intent);
        context.bindService(intent, sc, BIND_AUTO_CREATE);
    }

    private void addBookToService(ArrayList<Book> books) {
        int count = books.size();
        for (int i = 0; i < count; i++) {
            Book book = books.get(i);
            BookHelper.addDownBookTask(DownloadManagerActivity.this, book, views, true);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.title_back_btn:
                Map<String, String> data = new HashMap<>();
                data.put("type", "1");
                StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.SYSTEM_PAGE, StartLogClickUtil.BACK, data);
                finish();
                break;
            case R.id.title_name_btn:
                finish();
                break;
            case R.id.btn_edit:
                if (isDoubleClick(System.currentTimeMillis())) {
                    return;
                }
                if ("编辑".equals(editBtn.getText())) {
                    views.showRemoveMenu(views);
                    editBtn.setText("取消");
                    StatServiceUtils.statAppBtnClick(this, StatServiceUtils.bs_down_m_click_edit);
                    StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.CACHEEDIT1);
                } else {
                    views.dissmissremoveMenu();
                    editBtn.setText("编辑");
                    StatServiceUtils.statAppBtnClick(this, StatServiceUtils.bs_down_m_click_cancel);
                    StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CHCHEEDIT_PAGE, StartLogClickUtil.CANCLE);
                }

                break;
            default:
                break;
        }
    }

    @Override
    public boolean onMenuOpened(int featureId, Menu menu) {
        views.showRemoveMenu(views);
        return false;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (alertDialog != null && alertDialog.isShowing()) {
                alertDialog.cancel();

                return true;
            }
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
        if (!AppUtils.getPackageName().equals("cc.kdqbxs.reader")) {
            if (!views.isRemoveMode()) {
                views.showRemoveMenu(findViewById(R.id.root));
                editBtn.setText("取消");
            }
        }

        return false;
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        if (downloadingBooks == null || position < 0 || position > downloadingBooks.size()) {
            return;
        }
        if (position > downloadingBooks.size()) {
            return;
        }
        if (!views.isRemoveMode() && !views.isShowing) {
            Book book = downloadingBooks.get(position);
            Book b = (Book) mBookDaoHelper.getBook(book.book_id, 0);
            Map<String, String> data = new HashMap<>();
            data.put("STATUS", BookHelper.isDownFnish(this, b) ? "1" : "0");
            StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.BOOKCLICK1, data);
            BookHelper.goToCoverOrRead(getApplicationContext(), DownloadManagerActivity.this, b);
        } else {
            views.setRemoveChecked(position);
        }
    }

    @Override
    public void onBackPressed() {
        if (views.removehelper != null && views.removehelper.isRemoveMode()) {
            views.removehelper.dismissRemoveMenu();
            editBtn.setText("编辑");
        } else {
            //如果是从通知栏过来, 且已经退出到home了, 要回到应用中
            if (isTaskRoot()) {
                Intent intent = new Intent(this, SplashActivity.class);
                startActivity(intent);
            }
            super.onBackPressed();
        }
    }

    @Override
    public void notificationCallBack(Notification preNTF, String book_id) {
        PendingIntent pending;
        Intent intent;
        if (!book_id.equals(-1 + "")) {
            intent = new Intent(this, DownBookClickReceiver.class);
            intent.setAction(DownBookClickReceiver.action);
            intent.putExtra("book_id", book_id);
            pending = PendingIntent.getBroadcast(getApplicationContext(), 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        } else {
            intent = new Intent(this, DownloadManagerActivity.class);
            pending = PendingIntent.getActivity(getApplicationContext(), 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
        }
        preNTF.contentIntent = pending;
    }

    public boolean isDoubleClick(long time) {
        long length = time - lastClickTime;
        if (length > 800) {
            lastClickTime = time;
            return false;
        } else {
            return true;
        }
    }

}

File: common_book/src/mfqbxssc/java/com/intelligent/reader/activity/DownloadManagerActivity.kt
package com.intelligent.reader.activity

import com.intelligent.reader.R
import com.intelligent.reader.read.help.BookHelper
import com.intelligent.reader.view.DownloadPager
import com.intelligent.reader.view.DownloadPager.DeleteItemListener

import net.lzbook.kit.appender_loghub.StartLogClickUtil
import net.lzbook.kit.book.download.CacheManager
import net.lzbook.kit.book.download.DownloadState
import net.lzbook.kit.book.view.MyDialog
import net.lzbook.kit.constants.Constants
import net.lzbook.kit.data.bean.Book
import net.lzbook.kit.data.db.BookDaoHelper
import net.lzbook.kit.utils.FrameBookHelper.CachedComparator
import net.lzbook.kit.utils.FrameBookHelper.MultiComparator
import net.lzbook.kit.utils.SettingItemsHelper
import net.lzbook.kit.utils.StatServiceUtils

import android.content.Intent
import android.content.res.Resources.NotFoundException
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.v4.view.ViewCompat
import android.view.KeyEvent
import android.view.Menu
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemLongClickListener
import android.widget.ImageView
import android.widget.PopupWindow
import android.widget.TextView
import net.lzbook.kit.book.download.CallBackDownload

import java.util.ArrayList
import java.util.Collections
import java.util.HashMap

class DownloadManagerActivity : BaseCacheableActivity(), OnClickListener, OnItemClickListener, OnItemLongClickListener {
    private var alertDialog: MyDialog? = null
    private var back_btn: ImageView? = null
    private var content_layout: ViewGroup? = null
    private var downloadingBooks: ArrayList<Book>? = null
    var lastClickTime: Long = 0
    private var mBookDaoHelper: BookDaoHelper? = null
    private var more: ImageView? = null
    private var title_name_btn: TextView? = null
    private var editCancel: TextView? = null
    private var editSelectAll: TextView? = null
    var views: DownloadPager? = null

    override
    fun onCreate(paramBundle: Bundle?) {
        super.onCreate(paramBundle)
        try {
            setContentView(R.layout.download_manager)
        } catch (e: NotFoundException) {
            e.printStackTrace()
        }

        initView()
        initData()
        initListener()
        Constants.isDownloadManagerActivity = true
        Constants.hadShownMobilNetworkConfirm = false
    }

    override
    protected fun onStart() {
        super.onStart()
        initService()
    }

    override
    protected fun onStop() {
        super.onStop()
    }

    override
    protected fun onDestroy() {
        Constants.isDownloadManagerActivity = false
        Constants.hadShownMobilNetworkConfirm = false
        CacheManager.listeners.remove(this.views as CallBackDownload)
        if (this.alertDialog != null) {
            this.alertDialog = null
        }
        try {
            setContentView(R.layout.empty)
        } catch (e: NotFoundException) {
            e.printStackTrace()
        }

        super.onDestroy()
    }

    override
    protected fun onResume() {
        try {
            super.onResume()
            freshBooks(false)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun initService() {}

    fun stopDownloadbook(book_id: String) {
        CacheManager.stop(book_id)
    }

    private fun initView() {
        this.back_btn = findViewById(R.id.title_back_btn) as ImageView
        this.more = findViewById(R.id.ibtn_more) as ImageView
        this.content_layout = findViewById(R.id.content_layout) as ViewGroup
        this.title_name_btn = findViewById(R.id.title_name_btn) as TextView
        this.editCancel = findViewById(R.id.edit_cancel) as TextView
        this.editSelectAll = findViewById(R.id.edit_select_all) as TextView
    }

    private fun initListener() {
        this.back_btn!!.setOnClickListener(this)
        if (this.more != null) {
            this.more!!.setOnClickListener(this)
        }
        if (this.editCancel != null) {
            this.editCancel!!.setOnClickListener(this)
        }
        if (this.editSelectAll != null) {
            this.editSelectAll!!.setOnClickListener(this)
        }
        this.title_name_btn!!.setOnClickListener(this)
        this.views!!.setDeleteItemListener(object :DeleteItemListener {
            override
            fun onSuccess() {
                this@DownloadManagerActivity.editCancel!!.setVisibility(View.GONE)
                this@DownloadManagerActivity.editSelectAll!!.setVisibility(View.GONE)
                this@DownloadManagerActivity.more!!.setVisibility(View.VISIBLE)
                this@DownloadManagerActivity.back_btn!!.setVisibility(View.VISIBLE)
            }

            override
            fun onFailed() {
                this@DownloadManagerActivity.editCancel!!.setVisibility(View.GONE)
                this@DownloadManagerActivity.editSelectAll!!.setVisibility(View.GONE)
                this@DownloadManagerActivity.more!!.setVisibility(View.VISIBLE)
                this@DownloadManagerActivity.back_btn!!.setVisibility(View.VISIBLE)
            }
        })
        this.views!!.getListView().setOnItemClickListener(this)
        this.views!!.getListView().setOnItemLongClickListener(this)
    }

    private fun initData() {
        this.mBookDaoHelper = BookDaoHelper.getInstance()
        this.downloadingBooks = ArrayList()
        this.views = DownloadPager(getApplicationContext(), this, this.downloadingBooks)
        CacheManager.listeners.add(this.views as CallBackDownload)
        this.content_layout!!.addView(this.views)
    }

    fun freshBooks(hasDeleted: Boolean) {
        CacheManager.freshBooks(false)
        getDownLoadBookList(hasDeleted)
    }

    private fun getDownLoadBookList(hasDeleted: Boolean) {
        if (this.mBookDaoHelper == null) {
            this.mBookDaoHelper = BookDaoHelper.getInstance()
        }
        val books = this.mBookDaoHelper!!.getBooksOnLineList()
        if (this.downloadingBooks != null && this.views != null) {
            this.downloadingBooks!!.clear()
            this.downloadingBooks!!.addAll(books)
            Collections.sort(this.downloadingBooks)
            Collections.sort(this.downloadingBooks, MultiComparator())
            Collections.sort(this.downloadingBooks, CachedComparator())
            if (downloadingBooks!!.isEmpty()) {
                this.more!!.setVisibility(View.GONE)
            }
            this.views!!.freshBookList(this.downloadingBooks)
        }
    }

    fun checkSelectAll(isAll: Boolean) {
        if (isAll) {
            this.editSelectAll!!.setText("取消全选")
        } else {
            this.editSelectAll!!.setText("全选")
        }
    }

    override
    fun onClick(v: View) {

        val data = HashMap<String, String>()
        when (v.getId()) {
        //            case R.id.btn_edit:
        //                this.views.dissmissremoveMenu();
        //                this.editCancel.setVisibility(View.GONE);
        //                this.editSelectAll.setVisibility(View.GONE);
        //                this.more.setVisibility(View.VISIBLE);
        //                StatServiceUtils.statAppBtnClick(this, StatServiceUtils.bs_down_m_click_cancel);
        //                StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.CANCLE);
        //                return;
            R.id.edit_cancel -> {
                if (!isDoubleClick(System.currentTimeMillis())) {
                    this.views!!.removehelper.dismissRemoveMenu()
                    this.editCancel!!.setVisibility(View.GONE)
                    this.editSelectAll!!.setVisibility(View.GONE)
                    this.more!!.setVisibility(View.VISIBLE)
                    this.back_btn!!.setVisibility(View.VISIBLE)
                }
                return
            }
            R.id.edit_select_all -> {
                val isAll = "全选".equals(this.editSelectAll!!.getText())
                if (isAll) {
                    this.editSelectAll!!.setText("取消全选")
                } else {
                    this.editSelectAll!!.setText("全选")
                }
                views!!.setSelectAll(isAll)
                return
            }
            R.id.title_back_btn -> {
                data.put("type", "1")
                StartLogClickUtil.upLoadEventLog(this, "CACHEMANAGE", StartLogClickUtil.BACK, data)
                finish()
                return
            }
            R.id.title_name_btn -> {
                finish()
                return
            }
        //            case R.id.btn_edit:
        //                if (!isDoubleClick(System.currentTimeMillis())) {
        //                    if ("编辑".equals(this.editBtn.getText())) {
        //                        this.views.showRemoveMenu(this.views);
        //                        this.editBtn.setText("取消");
        //                        StatServiceUtils.statAppBtnClick(this, StatServiceUtils.bs_down_m_click_edit);
        //                        StartLogClickUtil.upLoadEventLog(this, "CACHEMANAGE", "CACHEEDIT");
        //                        return;
        //                    }
        //                    this.views.dissmissremoveMenu();
        //                    this.editBtn.setText("编辑");
        //                    StatServiceUtils.statAppBtnClick(this, StatServiceUtils.bs_down_m_click_cancel);
        //                    StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CHCHEEDIT_PAGE, "CANCLE");
        //                    return;
        //                }
        //                return;
            R.id.ibtn_more -> {

                StartLogClickUtil.upLoadEventLog(this, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.MORE)


                if (!isDoubleClick(System.currentTimeMillis())) {
                    val inflate = getLayoutInflater().inflate(R.layout.download_head_more_pop, null)
                    inflate.measure(0, 0)
                    val popupWindow = PopupWindow(inflate, -2, -2)
                    popupWindow.setBackgroundDrawable(ColorDrawable(ViewCompat.MEASURED_SIZE_MASK))
                    popupWindow.setFocusable(true)
                    popupWindow.setTouchable(true)
                    popupWindow.setOutsideTouchable(false)
                    this.more!!.getLocationOnScreen(IntArray(2))
                    popupWindow.showAsDropDown(this.more, 0, -this.more!!.getHeight())
                    val settingItemsHelper = SettingItemsHelper.getSettingHelper(getApplicationContext())
                    val txt_read = inflate.findViewById(R.id.download_head_pop_sort_read) as TextView
                    val txt_time = inflate.findViewById(R.id.download_head_pop_sort_time) as TextView
                    if (settingItemsHelper.getValues().booklist_sort_type === 0) {
                        txt_read.setTextColor(getResources().getColor(R.color.color_primary))
                    } else {
                        txt_time.setTextColor(getResources().getColor(R.color.color_primary))
                    }
                    txt_read.setOnClickListener(object : OnClickListener {
                        override
                        fun onClick(v: View) {
                            settingItemsHelper.putInt(settingItemsHelper.booklistSortType, 0)
                            Constants.book_list_sort_type = 0
                            popupWindow.dismiss()
                            this@DownloadManagerActivity.freshBooks(false)
                            data.put("type", "0")
                            StartLogClickUtil.upLoadEventLog(this@DownloadManagerActivity, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.SORT, data)
                        }
                    })
                    txt_time.setOnClickListener(object : OnClickListener {
                        override
                        fun onClick(v: View) {
                            settingItemsHelper.putInt(settingItemsHelper.booklistSortType, 1)
                            Constants.book_list_sort_type = 1
                            popupWindow.dismiss()
                            this@DownloadManagerActivity.freshBooks(false)
                            data.put("type", "1")
                            StartLogClickUtil.upLoadEventLog(this@DownloadManagerActivity, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.SORT, data)

                        }
                    })
                    inflate.findViewById(R.id.download_head_pop_edit).setOnClickListener(object : OnClickListener {
                        override
                        fun onClick(v: View) {
                            this@DownloadManagerActivity.editCancel!!.setVisibility(View.VISIBLE)
                            this@DownloadManagerActivity.editSelectAll!!.setVisibility(View.VISIBLE)
                            this@DownloadManagerActivity.more!!.setVisibility(View.GONE)
                            this@DownloadManagerActivity.back_btn!!.setVisibility(View.GONE)
                            this@DownloadManagerActivity.views!!.showRemoveMenu(this@DownloadManagerActivity.views)
                            StatServiceUtils.statAppBtnClick(this@DownloadManagerActivity, StatServiceUtils.bs_down_m_click_edit)
                            StartLogClickUtil.upLoadEventLog(this@DownloadManagerActivity, StartLogClickUtil.CACHEMANAGE_PAGE, StartLogClickUtil.CACHEEDIT)
                            popupWindow.dismiss()
                        }
                    })
                    return
                }
                return
            }
            else -> return
        }
    }

    override
    fun onMenuOpened(featureId: Int, menu: Menu): Boolean {
        this.editCancel!!.setVisibility(View.VISIBLE)
        this.editSelectAll!!.setVisibility(View.VISIBLE)
        this.more!!.setVisibility(View.GONE)
        this.back_btn!!.setVisibility(View.GONE)
        this.views!!.showRemoveMenu(this.views)
        return false
    }

    override
    fun onCreateOptionsMenu(menu: Menu): Boolean {
        return super.onCreateOptionsMenu(menu)
    }

    override
    fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode != 4 || this.alertDialog == null || !this.alertDialog!!.isShowing()) {
            return super.onKeyDown(keyCode, event)
        }
        this.alertDialog!!.cancel()
        return true
    }

    override
    fun onItemLongClick(adapterView: AdapterView<*>, view: View, position: Int, id: Long): Boolean {
        if (!this.views!!.isRemoveMode()) {
            this.editCancel!!.setVisibility(View.VISIBLE)
            this.editSelectAll!!.setVisibility(View.VISIBLE)
            this.more!!.setVisibility(View.GONE)
            this.back_btn!!.setVisibility(View.GONE)
            this.views!!.showRemoveMenu(findViewById(R.id.root))
        }
        return false
    }

    override
    fun onItemClick(adapterView: AdapterView<*>, view: View, position: Int, id: Long) {
        if (this.views!!.isRemoveMode() || this.views!!.isShowing) {
            this.views!!.setRemoveChecked(position)
            return
        }
        val book = this.views!!.getAdapter().getItem(position) as Book
        if (book != null) {
            val data = HashMap<String, String>()
            data.put("STATUS", if (CacheManager.getBookStatus(book) === DownloadState.FINISH) "1" else "0")
            StartLogClickUtil.upLoadEventLog(this, "CACHEMANAGE", "BOOKCLICK", data)
            BookHelper.goToCoverOrRead(getApplicationContext(), this, book, 1)
        }
    }

    override
    fun onBackPressed() {
        if (this.views!!.removehelper == null || !this.views!!.removehelper.isRemoveMode()) {
            if (isTaskRoot()) {
                startActivity(Intent(this, SplashActivity::class.java))
            }
            super.onBackPressed()
            return
        }
        this.views!!.removehelper.dismissRemoveMenu()
        this.editCancel!!.setVisibility(View.GONE)
        this.editSelectAll!!.setVisibility(View.GONE)
        this.more!!.setVisibility(View.VISIBLE)
        this.back_btn!!.setVisibility(View.VISIBLE)
    }

    fun isDoubleClick(time: Long): Boolean {
        if (time - this.lastClickTime <= 800) {
            return true
        }
        this.lastClickTime = time
        return false
    }

    override
    fun supportSlideBack(): Boolean {
        return if (isTaskRoot() || this.views!!.isRemoveMode()) false else true
    }

    companion object {
        private val TAG = "DownloadManagerActivity"
    }
}
